### PR TITLE
PP-3646 Implement `state` search and fix pagination

### PR DESCRIPTION
**Pull request by DanailMinchev**
## WHAT YOU DID

- add sql generation for `state` using mapping from `ExternalPaymentState`
- handle external states from resource level downwards
- fix pagination (it's page and not page_number)
- fix bug in calculating next/last/prev page
- update docs
- add/update relevant tests
- fix pagination url (no & after ?)
- added unit tests for `ViewPaginationBuilder`
- refactor `ViewPaginationBuilder` (remove unnecessary list of PaymentViewResultResponse

with @SandorArpa

### CODE REVIEW

File: src/main/java/uk/gov/pay/directdebit/payments/model/ViewPaginationBuilder.java [hunk 1]
```diff
@@ -80,10 +84,13 @@ private void buildLinks(long lastPage) {
         lastLink = PaginationLink.ofValue(uriWithParams(searchParams.buildQueryParamString()).toString());
 
         searchParams.withPage(selfPageNum - 1);
-        prevLink = selfPageNum == 1L ? null : PaginationLink.ofValue(uriWithParams(searchParams.buildQueryParamString()).toString());;
+        prevLink = selfPageNum == 1L ? null :
```
Comment (simad) [inline src/main/java/uk/gov/pay/directdebit/payments/model/ViewPaginationBuilder.java]: there is a fairly amount of logic here and I'd like to see it tested. I know it is implicitly tested elsewhere (like in `PaymentViewServiceTest.java`) but with unit tests here you could check all the paths without having to actually mess with fixtures and so on

File: src/main/java/uk/gov/pay/directdebit/payments/params/PaymentViewSearchParams.java [hunk 2]
```diff
@@ -29,15 +39,32 @@
...
+    private String state;
     private SearchDateParams searchDateParams;
     private PaginationParams paginationParams;
     private Map<String, Object> queryMap;
-    
+
+    private static Map<String, String> externalPaymentToInternalStateQueryMap = new HashMap<>(4);
+
+    static {
+        String andStateCondition = " AND t.state = '%s'";
```
Comment (simad) [inline src/main/java/uk/gov/pay/directdebit/payments/params/PaymentViewSearchParams.java]: I find this `andStateCondition` here not really readable. I would much rather put this in the StringBuilder below, see other comment.
The map would be simply: 

```
externalPaymentToInternalStateQueryMap
     .put(ExternalPaymentState.EXTERNAL_STARTED.getState(), NEW);
etc.
```

File: src/main/java/uk/gov/pay/directdebit/payments/params/PaymentViewSearchParams.java [hunk 3]
```diff
@@ -137,6 +171,9 @@ public String generateQuery() {
         if (amount != null) {
             sb.append(" AND t.amount = :" + AMOUNT_FIELD);
         }
+        if (isNotBlank(state) && externalPaymentToInternalStateQueryMap.containsKey(state)) {
+            sb.append(externalPaymentToInternalStateQueryMap.get(state));
```
Comment (simad) [inline src/main/java/uk/gov/pay/directdebit/payments/params/PaymentViewSearchParams.java]: here I would do: 
`sb.append(format(" AND t.state = '%s'", externalPaymentToInternalStateQueryMap.get(state)));`

I think it is clearer what's going on, and it is similar to the other fields

File: src/main/java/uk/gov/pay/directdebit/payments/params/PaymentViewSearchParams.java (hunk 2 above)
Comment (DanailMinchev) [reply src/main/java/uk/gov/pay/directdebit/payments/params/PaymentViewSearchParams.java]: Great suggestion! :) Will change that, not sure how to handle the case with https://github.com/alphagov/pay-direct-debit-connector/pull/223/files#diff-1f282af1a2bd8e2a97f7d49673214372R60 ?

Comment (simad) [reply src/main/java/uk/gov/pay/directdebit/payments/params/PaymentViewSearchParams.java]: We have clarified this via message.

File: src/main/java/uk/gov/pay/directdebit/payments/model/ViewPaginationBuilder.java (hunk 1 above)
Comment (DanailMinchev) [reply src/main/java/uk/gov/pay/directdebit/payments/model/ViewPaginationBuilder.java]: Added in https://github.com/alphagov/pay-direct-debit-connector/pull/223/commits/c53d425f52adf6d6c6b3b223d5ce71bb99ab5760

**Pull request merged**